Reimplementation of glob in Crystal

pgkos opened the pull request:
This is an attempt to reimplement glob in Crystal. I have tried to make it behave exactly like in Ruby, therefore it also supports recursive glob. So, it should fix #1433.

It works by converting the glob pattern into a regex, traversing the directory recursively and matching the relative paths.

There are still some issues which we have to consider:
* should we recurse into paths pointed to by symbolic links?
* should the file read/access errors be raised (now they are rescued silently)?
* improve and test the glob's syntax on Windows (of course after implementing the other methods using Windows API)
* should there be an instance glob method? (so it is not required to use `chdir`)

Sorry for the wordy `scandir` method. Please consider supporting recursive `yield` (it would simplify the code very much)!

jhass commented:
Thank you!

I think we will need quite a few specs for this, however using the repo has fixture should be acceptable, we can update the tests as it changes.

jhass commented:
Also please replace all existing usages of `LibC.glob` and drop it.

jhass commented on src/dir.cr:
```diff
@@ -255,29 +255,147 @@ class Dir
+    level_stack = [level]
+    dir_stack = [] of Dir
+    recurse = true
+    while !dir_path_stack.empty?
```
`until dir_path_stack.empty?`

jhass commented on src/dir.cr:
```diff
@@ -255,29 +255,147 @@ class Dir
-
-    Slice(UInt8*).new(paths.pathv, paths.pathc.to_i32).each do |path|
-      yield String.new(path)
+    return Regex.new("^#{regex_pattern}$")
```
Should probably use `\A` and `\z` instead of `^` and `$` respectively, just to be on the safe side.

jhass commented on src/dir.cr:
```diff
@@ -255,29 +255,147 @@ class Dir
+    # Windows paths must have double backslashes:
+    escapable = ['?', '{', '}', '*', ',', '\\']
+    # characters which must be escaped in a PCRE regex:
+    escaped = ['$', '(', ')', '+', '.', '[', '^', '|', '/']
```
These two should be tuples `{'?', '{', ...}`

waj replied:
And constants?

jhass replied:
I guess so, yeah. Makes me wish we had private constants :P

asterite replied:
Why move them to constants if they are immutable?

jhass replied:
Basically because they're immutable, so you don't accidentally reassign or something.

jhass commented on src/dir.cr:
```diff
@@ -255,29 +255,147 @@ class Dir
+        rel_path = rel_path_stack[-1]
+        level = level_stack[-1]
+        dir = dir_stack[-1]
+      end
```
I'd replace these `[-1]` with more expressive `.last` calls.

jhass commented:
Mmh, I think we indeed do should recurse into symlinks.

technorama commented:
Following symlinks could send you in an infinite loop.

Ruby doesn't follow symlinks without weird arguments.
https://stackoverflow.com/questions/357754/can-i-traverse-symlinked-directories-in-ruby-with-a-glob

pgkos commented:
Yes, I knew about that, you need to store a list/hash set of already visited paths.

pgkos commented:
The specs are failing because my implementation always globs relatively to (inside) current working directory, but the specs expect the glob to treat the pattern as if it represents absolute path.

Should I then add support for absolute-path globbing? (maybe we should just treat glob patterns beginning with `/` as absolute?)

jhass commented:
Yes, that sounds completely reasonable

jhass commented:
Okay, I think this is good enough for a first implementation and we iterate on it smaller follow up PRs for things like symlink handling or reducing the amount of `stat` calls etc. Globbing is known to be a fairly slow operation so I think it's okay if it's not highly optimized from the start.

The one thing I'd still want to see included here however are some specs, especially around patterns including `**` but also using alternatives (`{a,b}`) and the combination of both for good measure.

pgkos commented:
Okay, I will try to write some new specs.

Also, I have some ideas on optimizing the glob (e.g. not recursing needlessly into some directories, so the glob search begins with the directory which is as nested as possible), but maybe I will leave this for later PRs.

asterite commented:
@pgkos Thanks for this! This is really needed  :-)

I tried this:

```crystal
puts Dir["**/*"].join("\n")
```

But I don't get the same results as in Ruby. It seems the Crystal version doesn't go inside subdirectories.

pgkos commented:
@asterite you mean my new implementation? I have just checked `puts Dir["**/*"]` and got exactly the same results as in Ruby.

asterite commented:
@pgkos Um, so sorry. I tested this thinking it was merged, then commented here... but I see it's not merged :blush:

pgkos commented:
@asterite not yet, I need to write specs for the new glob syntax :smile:

pgkos commented:
I have added two additional tests for recursive glob. Also, I have fixed some bugs and implemented an optimization which dramatically reduces time for deeply nested globs (mainly when using absolute paths).

jhass commented:
Looking good! Though given the complex pattern transformation, let's think of a couple more edge cases. Let's see

* `{spec/std,src}/**/*.cr`
* `src/file{/*.cr,.cr}`
* `spec/{**/*_spec,spec_helper}.cr`
* Can we actually nest alternations? `src/i{?,{terable,iterator}}.cr`

How does that look?

pgkos commented:
Your examples should work - the glob pattern gets converted into a regex, so in your last example:
```
rc/i{?,{terable,iterator}}.cr
```
is converted into:
```
rc/i(?:.|(?:terable|iterator)).cr
```
The curly braces are converted into non-capturing groups, `?`s are converted into `.` and `,` are converted (only if inside curly braces) into `|`.
Also:
* `*` are converted into `[^\\/]*`
* `**` into `.*` and
* `**/` into `(?:.*/)`

jhass commented:
That's good to hear, I'd just really enjoy specs that ensure they'll continue to work in the future too ;) I tried to pick patterns actually yielding results for the current source tree.

pgkos commented:
Okay, I'll put them into specs. Let's hope they will work:smile:

pgkos commented:
I think the tests fail because I don't know the *absolute path* (`__DIR__`, used by Travis) to the Crystal source (is it `/home/travis/build/manastech/crystal`? should I hard-code it into the tests?).

I cannot use relative globs (globs which don't start with a `/`) because the working dir is inside `/src`, so I cannot write a glob which will recurse inside e.g. `/spec`.

jhass commented:
Mmh, no they absolutely shouldn't hardcode the absolute path. Can't you use `..` in front?

pgkos commented:
Sure

jhass commented:
Okay, the last thing I'd like to ask from you is to squash everything into a commit, then we really should be good :)

```sh
git remote add upstream https://github.com/manastech/crystal.git
git fetch upstream
git checkout
git rebase -i upstream/master
# Leave the first commit at pick, choose squash or fix for the others.
# If there are any conflicts git will inform you about that, resolve them for
# example with git mergetool or manually and with git add, then run
# git rebase --continue, again  watch the output since that might happen for
# each commit.
git log # Look at the updated commit history
git status # should report you're back on your branch
# We need to force push since a rebase edits the commits, effectively
# creating new ones. This is okay to do in a feature branch, the pull
# request will update.
git push -f origin # Assuming you cloned from your fork
```

Hint: You can also edit your last commit with `git commit --amend`, since that also effectively creates a new commit, you to need to force push afterwards and again you should only do it in feature branches.

jhass commented:
Well, after fixing the four new specs of course ;)

pgkos commented:
I have squashed the commits. All the test cases are working.

Pathological patterns, like:
```
/foo/bar/*ar/../bar
```
won't work. The `..` directory, if present in the glob pattern, must always be before the first special character, so this works:
```
../foo/bar/*ar/bar
```

jhass commented:
:heart: thank you very much!
